connectors/sources/box.py

```diff
@@ -0,0 +1,445 @@
+                }
+                async with self._http_session.post(
+                    url=ENDPOINTS["TOKEN"],
+                    data=data,
+                ) as response:
+                    tokens = await response.json()
+                    self.access_token = tokens.get("access_token")
+                    refresh_token = tokens.get("refresh_token")
+                    self.expired_at = datetime.utcnow() + timedelta(
+                        seconds=int(tokens.get("expires_in", 3599))
+                    )
+                    self._token_cache.set_value(
+                        value=self.access_token, expiration_date=self.expired_at
+                    )
+            else:
+                data = {
+                    "client_id": self.client_id,
+                    "client_secret": self.client_secret,
+                    "grant_type": "client_credentials",
+                    "box_subject_type": "enterprise",
+                    "box_subject_id": self.enterprise_id,
+                }
+                async with self._http_session.post(
+                    url=ENDPOINTS["TOKEN"],
+                    data=data,
+                ) as response:
+                    tokens = await response.json()
+                    self.access_token = tokens.get("access_token")
+                    self.expired_at = datetime.utcnow() + timedelta(
+                        seconds=int(tokens.get("expires_in", 3599))
+                    )
+                    self._token_cache.set_value(
+                        value=self.access_token, expiration_date=self.expired_at
+                    )
+        except Exception as exception:
+            raise TokenError(
+                f"Error while generating access token. Please verify that provided configurations are correct. Exception {exception}."
+            ) from exception
+
+
+class BoxClient:
+    def __init__(self, configuration):
+        self._sleeps = CancellableSleeps()
+        self.configuration = configuration
+        self._logger = logger
+        self.is_enterprise = configuration["is_enterprise"]
+        self._http_session = aiohttp.ClientSession(
+            base_url="https://api.box.com", raise_for_status=True
+        )
+        self.token = AccessToken(
+            configuration=configuration, http_session=self._http_session
+        )
+
+    def set_logger(self, logger_):
+        self._logger = logger_
+
+    async def _put_to_sleep(self, exception):
+        retry_after = int(exception.headers.get("retry-after", 5))
+        self._logger.debug(
+            f"Connector will attempt to retry after {retry_after} seconds."
+        )
+        await self._sleeps.sleep(retry_after)
+        raise Exception("Rate limit exceeded.")
+
+    async def _handle_client_errors(self, exception):
+        match exception.status:
+            case 401:
+                await self.token._set_access_token()
+                raise
+            case 429:
```

Comment:
What's the purpose of using the custom `_put_to_sleep` method instead of relying on  `@retryable` wrapper?

Reply:
We are using `_put_to_sleep` because at the time of `429` we need to wait for some specific amount of time, which we get from the retry-after header, but how can we pass it to `@retryable`?

Reply:
Alright, there is no way to pass it to @retryable right now.